### PR TITLE
Add snap pipeline

### PR DESCRIPTION
**Pull request by matkt**
## PR description

This PR adds the majority of  classes related to the snapsync pipeline. More tests will be added later.
There will be a dedicated PR to delete hard values in order to add configuration flags 

## Documentation

- [ ] I thought about documentation and added the `doc-change-required` label to this PR if
    [updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).

## Changelog

- [ ] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/LoadLocalDataStep.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    // check if node is already stored in the worldstate
+    if (snapSyncState.hasPivotBlockHeader()) {
+      Optional<Bytes> existingData = request.getExistingData(worldStateStorage);
+      if (existingData.isPresent()) {
+        existingNodeCounter.inc();
+        request.setData(existingData.get());
+        request.setRequiresPersisting(false);
+        request.setRootHash(snapSyncState.getPivotBlockHeader().get().getStateRoot());
+        final WorldStateStorage.Updater updater = worldStateStorage.updater();
+        request.persist(worldStateStorage, updater, downloadState, snapSyncState);
```
Comment (fab-10) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/LoadLocalDataStep.java:65]: question: if this data already exists, why we persist it again?

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/StubTask.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright ConsenSys AG.
```
Comment (fab-10) [inline ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/StubTask.java:2]: wrong copyright

Comment (matkt) [reply ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/StubTask.java:2]: still this copyright issues 😵‍💫

Comment (matkt) [reply ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/StubTask.java:2]: done

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/DynamicPivotBlockManager.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
+package org.hyperledger.besu.ethereum.eth.sync.snapsync;
```
[the remark this answers is not in the document]
Comment (matkt) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/DynamicPivotBlockManager.java:15]: done

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/DynamicPivotBlockManager.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+                          .orElse(currentPivotBlockNumber);
+              if (distanceNextPivotBlock > 60 && isSearchingPivotBlock.compareAndSet(false, true)) {
+                syncActions
+                    .waitForSuitablePeers(FastSyncState.EMPTY_SYNC_STATE)
+                    .thenCompose(syncActions::selectPivotBlock)
+                    .thenCompose(syncActions::downloadPivotBlockHeader)
+                    .thenAccept(fss -> lastBlockFound = fss.getPivotBlockHeader())
+                    .orTimeout(5, TimeUnit.MINUTES)
+                    .whenComplete((unused, throwable) -> isSearchingPivotBlock.set(false));
+              }
```
[the remark this answers is not in the document]
Comment (matkt) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/DynamicPivotBlockManager.java:76]: I removed the duplication for SnapSyncDownloader and FastSyncDownloader. On this class it's not really the same behavior. So I prefer to keep it like that

**Review by fab-10 — APPROVED** [review]

**Pull request merged**